controllers/topology/internal/extensions/patches/inline/json_patch_generator.go

```diff
@@ -0,0 +1,368 @@
+	value := apiextensionsv1.JSON{}
+	if err := yaml.Unmarshal(buf.Bytes(), &value); err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal rendered template: %q", buf.String())
+	}
+
+	return &value, nil
+}
+
+// calculateTemplateData calculates data for the template, by converting
+// the variables to their Go types.
+// Example:
+// * Input:
+//   map[string]apiextensionsv1.JSON{
+//     "builtin": {Raw: []byte(`{"cluster":{"name":"cluster-name"}}`},
+//     "integerVariable": {Raw: []byte("4")},
+//     "numberVariable": {Raw: []byte("2.5")},
+//     "booleanVariable": {Raw: []byte("true")},
+//   }
+// * Output:
+//   map[string]interface{}{
+//     "builtin": map[string]interface{}{
+//       "cluster": map[string]interface{}{
+//         "name": <string>"cluster-name"
+//       }
+//     },
+//     "integerVariable": <float64>4,
+//     "numberVariable": <float64>2.5,
+//     "booleanVariable": <bool>true,
+//   }
+func calculateTemplateData(variables map[string]apiextensionsv1.JSON) (map[string]interface{}, error) {
+	res := make(map[string]interface{}, len(variables))
+
+	// Marshal the variables into a byte array.
+	tmp, err := json.Marshal(variables)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to convert variables: failed to marshal variables")
+	}
+
+	// Unmarshal the byte array back.
+	// NOTE: This converts the "leaf nodes" of the nested map
+	// from apiextensionsv1.JSON to their Go types.
+	if err := json.Unmarshal(tmp, &res); err != nil {
+		return nil, errors.Wrapf(err, "failed to convert variables: failed to unmarshal variables")
+	}
+
+	return res, nil
+}
+
+// mergeVariableMaps merges variables.
+// NOTE: In case a variable exists in multiple maps, the variable from the latter map is preserved.
+// NOTE: The builtin variable object is overwritten instead of simply overwritten.
+func mergeVariableMaps(variableMaps ...map[string]apiextensionsv1.JSON) (map[string]apiextensionsv1.JSON, error) {
+	res := make(map[string]apiextensionsv1.JSON)
+
+	for _, variableMap := range variableMaps {
+		for variableName, variableValue := range variableMap {
+			// If the variable already exits and is the builtin variable, merge it.
+			if _, ok := res[variableName]; ok && variableName == api.BuiltinVariableName {
+				mergedV, err := mergeBuiltinVariables(res[variableName], variableValue)
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed to merge builtin variables")
+				}
+				res[variableName] = *mergedV
+				continue
+			}
+			res[variableName] = variableValue
+		}
+	}
+
+	return res, nil
+}
+
+// mergeBuiltinVariables merges builtin variable objects.
+// NOTE: In case a variable exists in multiple builtin variables, the variable from the latter map is preserved.
+func mergeBuiltinVariables(variableList ...apiextensionsv1.JSON) (*apiextensionsv1.JSON, error) {
+	builtinVariable := &api.BuiltinVariable{}
+
+	// Unmarshal all variables into builtinVariable.
```

Comment:
Q: does this keep accumulating values given that we are marshalling on top of the same object?
If yes this should be better documented because this is not immediately visible

Reply:
Yes and yes I'll add more text.

I went with this solution to avoid writing a "merge nested maps" func.

Reply:
Extended the comment